src/k2/persistence/logStream/LogStream.cpp

```diff
@@ -0,0 +1,287 @@
+        throw std::runtime_error("Log stream already created");
+    }
+
+    _logStreamName = std::move(logStreamName);
+    _logStreamCreate = true;
+
+    // create metadata plogs in advance
+    std::vector<seastar::future<std::tuple<Status, String> > > createFutures;
+    for (uint32_t i = 0; i < METADATA_PLOG_POOL_SIZE; ++i){
+        createFutures.push_back(_client.create());
+    }
+    return seastar::when_all_succeed(createFutures.begin(), createFutures.end())
+    .then([this] (auto&& responses){
+        // register the metadata plog info to CPO
+        for (auto& response: responses){
+            auto& [status, plogId] = response;
+            if (!status.is2xxOK()){
+                throw std::runtime_error("unable to create plog for Metadata LogStream");
+            }
+            _metadataPlogPool.push_back(std::move(plogId));
+        }
+        String plogId = _metadataPlogPool[_metadataPlogPool.size()-1];
+        _metadataPlogPool.pop_back();
+        _plogInfo.push_back(std::make_pair(std::move(plogId), 0));
+        return _cpo.RegisterMetadataStreamLog(Deadline<>(_cpo_timeout()), _logStreamName, _plogInfo[0].first);
+    })
+    .then([this] (auto&& response){
+        // create WAL plogs in advance
+        auto& [status, resp] = response;
+        if (!status.is2xxOK()) {
+            throw std::runtime_error("unable to register Metadata LogStream Plog to CPO");
+        }
+        std::vector<seastar::future<std::tuple<Status, String> > > createFutures;
+        for (uint32_t i = 0; i < WAL_PLOG_POOL_SIZE; ++i){
+            createFutures.push_back(_client.create());
+        }
+        return seastar::when_all_succeed(createFutures.begin(), createFutures.end());
+    })
+    .then([this] (auto&& responses){
+        for (auto& response: responses){
+            auto& [status, plogId] = response;
+            if (!status.is2xxOK()){
+                throw std::runtime_error("unable to create plog for WAL");
+            }
+            _walPlogPool.push_back(std::move(plogId));
+        }
+        String plogId = _walPlogPool[_walPlogPool.size()-1];
+        _walPlogPool.pop_back();
+        _plogInfo.push_back(std::make_pair(plogId, 0));
+
+        // write the metadata of the first WAL plog to metadata log streams
+        Payload temp_payload([] { return Binary(4096); });
+        temp_payload.write(std::move(plogId));
+        return _write(std::move(temp_payload), 0);
+    });
+}
+
+seastar::future<std::vector<Payload> >
+LogStream::read(String logStreamName){
+    K2INFO("Received Logstream Read Request for " << logStreamName);
+    // read the metadata of metadata log stream from CPO
+    return _cpo.GetMetadataStreamLog(Deadline<>(_cpo_timeout()), logStreamName)
+    .then([&] (auto&& response){
+        auto& [status, resp] = response;
+        if (!status.is2xxOK()) {
+            throw std::runtime_error("unable to obtain metadata log stream");
+        }
+        _streamLog = std::move(resp.streamLog);

```

Comment:
Using a class variable here means it won't be safe for concurrent calls to read, it would be better to have it allocated on the heap.

Reply:
fixed